maggy/optimizer/bayes/base.py

```diff
@@ -0,0 +1,1016 @@
+                )
+                i += 1
+                if i > 3:
+                    self._log(
+                        "not possible to sample new config. Stop Experiment (most/all configs have already been used)"
+                    )
+                    return None
+
+            # report new trial id to pruner
+            if self.pruner:
+                self.pruner.report_trial(
+                    original_trial_id=None, new_trial_id=next_trial.trial_id
+                )
+            self._log(
+                "start trial {}: {}, {} \n".format(
+                    next_trial.trial_id, next_trial.params, next_trial.info_dict
+                )
+            )
+            return next_trial
+
+        except BaseException:
+            self._log(traceback.format_exc())
+            self._close_log()
+
+            if self.pruner:
+                self.pruner._close_log()
+
+    def finalize_experiment(self, trials):
+        self._log("Experiment finished")
+        self._close_log()
+
+        if self.pruner:
+            self.pruner._close_log()
+
+        return
+
+    def init_model(self):
+        """initializes the surrogate model of the gaussian process
+
+        the model gets created with the right parameters, but is not fit with any data yet. the `base_model` will be
+        cloned in `update_model` and fit with observation data
+        """
+        raise NotImplementedError
+
+    def init_pruner(self, pruner, pruner_kwargs):
+        """intializes pruner
+
+        :param pruner: name of pruner. so far only "hyperband" implemented
+        :type pruner: str
+        :param pruner_kwargs: dict of pruner kwargs
+        :type pruner_kwargs: dict
+        :return: initiated pruner instance
+        """
+        allowed_pruners = ["hyperband"]
+        if pruner not in allowed_pruners:
+            raise ValueError(
+                "expected pruner to be in {}, got {}".format(allowed_pruners, pruner)
+            )
+
+        if pruner == "hyperband":
+            self.pruner = Hyperband(
+                trial_metric_getter=self.get_metrics_dict, **pruner_kwargs
+            )
+
+    def update_model(self, budget=0):
+        """update surrogate model with new observations
+
+        Use observations of finished trials + liars from busy trials to build model.
+        Only build model when there are at least as many observations as hyperparameters
+
+        :param budget: the budget for which model should be updated. Default is 0
+                       If budget > 0 : multifidelity optimization. Only use observations that were run with
+                       `budget` for updateing model for that `budget`. One model exists per budget.
+                       If == 0: single fidelity optimization. Only one model exists that is fitted with all observations
+        :type budget: int
+        """
+        raise NotImplementedError
```

Comment:
same as above, can it be an abstractmethod?

Reply:
yes